pkg/tfshim/sdk-v2/provider2.go

```diff
@@ -798,3 +803,83 @@ func recoverDiagnostic(d tfprotov5.Diagnostic) diag.Diagnostic {
 	}
 	return dd
 }
+
+func normalizeBlockCollections(val cty.Value, res *schema.Resource) cty.Value {
+	// Full rules about block vs attr
+	//nolint:lll
+	// https://github.com/hashicorp/terraform-plugin-sdk/blob/1f499688ebd9420768f501d4ed622a51b2135ced/helper/schema/core_schema.go#L60
+	sch := res.CoreConfigSchema()
+	if !val.Type().IsObjectType() {
+		contract.Failf("normalizeBlockCollections: Expected object type, got %v", val.Type().GoString())
+	}
+	valMap := val.AsValueMap()
+
+	for fieldName := range sch.BlockTypes {
+		if !val.Type().HasAttribute(fieldName) {
+			continue
+		}
+		subVal := val.GetAttr(fieldName)
+		if subVal.IsNull() {
+			fieldType := val.Type().AttributeType(fieldName)
+			// Only lists and sets can be blocks and pass InternalValidate
+			// Ignore other types.
+			if fieldType.IsListType() {
+				glog.V(10).Info("normalizing list %s, %s", fieldName, fieldType.ElementType())
+				valMap[fieldName] = cty.ListValEmpty(fieldType.ElementType())
+			} else if fieldType.IsSetType() {
+				glog.V(10).Info("normalizing set %s, %s", fieldName, fieldType.ElementType())
+				valMap[fieldName] = cty.SetValEmpty(fieldType.ElementType())
+			}
+		} else {
+			subBlockSchema := res.SchemaMap()[fieldName]
+			if subBlockSchema == nil {
+				glog.V(5).Info("normalizeBlockCollections: Unexpected nil subBlockSchema for %s", fieldName)
+				continue
+			}
+			subBlockRes, ok := subBlockSchema.Elem.(*schema.Resource)
+			if !ok {
+				glog.V(5).Info("normalizeBlockCollections: Unexpected schema type %s", fieldName)
```

Comment:
This branch doesn't sound exceptional. I think this can be an attribute with Elem=Schema or something with Elem=nil, fairly normal no? If the hunch is right perhaps we should be silent in this case, not log anything.

Reply:
I believe it is exceptional since we are iterating over sch.BlockTypes and these should only contain the block types - this should mean Elem = Resource && Type == List || Set